plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction/Extract.hs

```diff
@@ -0,0 +1,268 @@
+--
+-- Original range:
+--            1 + 1
+--      ^               ^
+--
+-- UnpaddedRange:
+--            1 + 1
+--            ^   ^
+-- See `removeWhitespacePadding` for more info
+newtype UnpaddedRange = UnpaddedRange {unUnpaddedRange :: Range}
+
+-- | This function removes whitespace padding from the Range provided by the user.
+--
+-- When the user selects their range that they would like to extract, we would like to ignore whitespace in their
+-- selection, so that we can find expressions that are contained fully by said range. Consider the following selection:
+--
+--   foo = 1    +    2
+--           ^       ^
+-- In this case, while the range fully wraps `2`, it is not a valid extract function query, since it also wraps `+ 2`,
+-- which is not an expression.
+--
+-- Therefore, in order to catch this case, we narrow the selection to:
+--
+--   foo = 1    +    2
+--              ^    ^
+-- and look for LHsExprs that precisely match the given Range, which ensures that we only accept a range if:
+--  1. the range wraps an entire LHsExpr
+--  2. the range wraps an LHsExpr exactly
+--
+-- More explanation of (2)... If a range does exactly match any LHsExpr, then we know it is an invalid range for
+-- extract, since there is no whitespace padding in the range.
+removeWhitespacePadding :: Range -> T.Text -> UnpaddedRange
+removeWhitespacePadding range@(Range (Position sl sc) (Position el ec)) source =
+  let selectedTxt = textInRange range source
+      -- TODO We do not yet handle tabs gracefully
+      isWhiteSpace c = elem c [' ', '\n']
+      (paddingLeft, minimalOnLeft) = T.span isWhiteSpace selectedTxt
+      (paddingRight, _) = T.span isWhiteSpace $ T.reverse minimalOnLeft
+      whitespaceToDelta = \case
+        ' '  -> (0 :: Sum UInt, -1 :: Sum UInt)
+        '\n' -> (-1, 0)
+        _    -> error "impossible"
+      paddingToDelta = foldMap whitespaceToDelta . T.unpack
+      (getSum -> rowsL, getSum -> colsL) = paddingToDelta paddingLeft
+      (getSum -> rowsR, getSum -> colsR) = paddingToDelta paddingRight
+     in UnpaddedRange $ Range (Position (sl - rowsL) (sc - colsL)) (Position (el - rowsR) (ec - colsR))
+
+-- | Queries an AST for the topmost LHsExpr that exactly matches the given UnpaddedRange.
+tryFindLargestExprInRange :: UnpaddedRange -> GenericQ (Maybe (SrcSpan, LHsExpr GhcPs))
+tryFindLargestExprInRange range = everythingBut (<|>) $ mkQ (Nothing, False) firstContainedExprQ
+  where
+    -- When we find an LHsExpr that is fully contained by this range, we return the LHsExpr and early-exit the
+    -- syb traversal (by returning True)
+    firstContainedExprQ = \case
+      lexpr@(L (SrcSpanAnn _ span@(RealSrcSpan realSrcSpan _)) _) :: LHsExpr GhcPs
+        | rangeSrcSpan <- rangeToSrcSpan (fromString $ unpackFS $ srcSpanFile realSrcSpan) (unUnpaddedRange range)
+        , isSameSrcSpanModuloFile rangeSrcSpan span
```

Comment:
why modulo file? Maybe we should be turning the srcspans into Ranges and comparing those instead?